### ISSUE TITLE
some questions

### ISSUE THREAD

**Issue by dst21** [R0Wi/workflow_ocr#36]
hi,

I like this workflow and am testing different scenarios. I have two more questions about this:

1. Does the OCR also work in external directories?
2. If not, is a file considered new or changed if it is moved from an external directory to an internal one?

Background: I have a scanner that stores the PDFs on SMB shares. Now I would like to revise these PDFs through this workflow, because OCRmyPDF under Linux shell and Cron sometimes deletes data if they are stored while OCR is running.

Thanks in advance for the workflow and your efforts. ;)

**Comment by edd--** [comment]
same question. Workflow is not being triggered when using occ files:scan, any way 'round that ?

**Comment by R0Wi** [comment]
Well i'm also using the workflow on a samba share but i'm writing the files to it via NC (web or mobile app) so there it works that NC recognizes the new file an triggers the workflow.

To be honest i did not test if it works when i create the file on the samba share and trigger a `occ files:scan` afterwards so that NC takes note of the new file. But if you tell me it does not work the issue is most likely present inside of the NC core (or even the workflow engine itself) because currently the workflow is just sticked to the [`postWrite`](https://docs.nextcloud.com/server/17/developer_manual/app/hooks.html#filesystem-root) hook via basic workflow infrastructure. Let me investigate and i'll give you some feedback on this :-)

**Comment by edd--** [comment]
great thanks ! For now I created a workaround with a separate docker container that uploads the files in the FTP folder using webdav whenever a new file is detected, but that's kind of detour, and thus more error prone ...

Appreciate it !

**Comment by R0Wi** [comment]
Little update on this: i did a quick test by adding a file to the samba share directly and then i issued the `occ files:scan --all` command. My debugger did not hit the code where usually the file will be added to the process queue so i supect this to be an error in NC core. Digging deeper these days ...

**Comment by R0Wi** [comment]
Ok so i spent some time debugging and here are the results:

- If you use an external storage where you directly put your files to (without going through NC, so for example if you copy a file directly to a samba share) the `postWrite` and `postCreate` [hooks](https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/external_storage/smb.html#smb-update-notifications) are not fired. These events are associated with the `File updated` and `File modified` conditions in the workflow
![image](https://user-images.githubusercontent.com/19730957/99628709-fd3a7780-2a36-11eb-93cd-c051bada9729.png)
- If you use `occ files:scan` command this will **not** trigger the mentioned hooks but rather emit events like [`\OCP\Files\Events\NodeAddedToCache`](https://docs.nextcloud.com/server/latest/developer_manual/basics/events.html?highlight=hooks#ocp-files-events-nodeaddedtocache) 
- Unfortunately the events which occur while scanning the NC instance are currently not available or associated with any condition which can be chosen in the workflow conditions dropdown

So to summarize: i'm not sure if this use case is yet tackled by the workflow engine or if there are plans to support conditions like "`External file added`" or something. But this is definitely something which must be pinned to NC workflow engine itself at https://github.com/nextcloud/server. Hope this helps somehow :-)

**Comment by derlucas** [comment]
I got a quick question too: How to scan only files in a certain folder? I can use the "File Name" matches Filter, but i read that this does not apply to folder/paths and only to the actual file name.

**Comment by R0Wi** [comment]
This is indeed a feature i'm missing, too. But unfortunately the filter section is controlled by the workflow core engine inside of Nextcloud.

Might be worth to open a new issue there?

**Comment by sirweazel** [comment]
I'm running the nextcloud snap, should the flow OCR work in the nextcloud snap?  It installed, but didn't seem to get triggered.  I"m not sure yet if it is a flow issue or the ocr part.  Just checking before i spend some time investigating?  I know the snap is missing imagick if that is used i think.  I will also mention, as the other stated, the ability to control more which files or filter OCR is applied to would be beneficial, or define where the final created files end up at.  I'll try and find the the flow core engine github and open or look at their issues and add if i can.

**Comment by dst21** [comment]
I have currently solved it as follows so that my documents can be processed:

The scanner scans the SMB share on a TrueNAS file backup instance> this uploads the data to my Nextcloud via "cloud upload" and the workflow carries out OCR. With this I got around the problem that with an SMB share integrated in NC (where a local OCR script is running over it) files are sometimes deleted if they are uploaded from the scanner during a running OCR job controlled by CRON.

**Comment by R0Wi** [comment]
Closing this for now. Please request reopen if necessary.

**Comment by nicweb** [comment]
Is this still the status quo? I ran into the same issue with my setup (scan to smb - available as external folder for all local users).

**Comment by R0Wi** [comment]
@nicweb please see my answer here https://github.com/R0Wi/workflow_ocr/issues/36#issuecomment-730157406

> If you use an external storage where you directly put your files to (without going through NC, so for example if you copy a file directly to a samba share) the postWrite and postCreate [hooks] (https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/external_storage/smb.html#smb-update-notifications) are not fired. These events are associated with the File updated and File modified conditions in the workflow

Like stated in the [docs](https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/external_storage/smb.html#smb-update-notifications), when creating a file directly on SMB:

> Nextcloud can use smb update notifications to listen for changes made to a configured SMB/CIFS storage and detect external changes made to the storage in near real-time. **Note: Due to limitations of linux based SMB servers, this feature only works reliably on Windows SMB servers.**

Unfortunately, this is currently a limitation from NC itself and not directly related to the workflowengine nor this app 😢

**Comment by nicweb** [comment]
Thanks for the clarification -  I read your previous comment, but hoped for a change NC server side within the last 1.5 years.
Great app nevertheless - thanks for all your work.
